### PR TITLE
Rework multiple graphs implementation in frontend

### PR DESCRIPTION
**Pull request by Mats-SX**

### CODE REVIEW

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/semantics/SemanticFeature.scala [hunk 1]
```diff
@@ -21,4 +21,5 @@ sealed trait SemanticFeature
 object SemanticFeature {
   case object MultipleGraphs extends SemanticFeature
   case object WithInitialQuerySignature extends SemanticFeature
+  case object Cypher10Support extends SemanticFeature
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/semantics/SemanticFeature.scala:24]: Is there a need for `MultipleGraphs` if we have `Cypher10Support`?

File: community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala [hunk 2]
```diff
@@ -148,25 +148,38 @@ class InvalidNodePattern(
...
   override def isSingleNode = true
 }
 
+sealed trait EquivalenceModel extends ASTNode {
+  def variable: LogicalVariable
+}
+final case class TildeModel(variable: LogicalVariable)(val position: InputPosition) extends EquivalenceModel {
+  override def toString: String = "`~` (equivalence)"
+}
+final case class AtModel(variable: LogicalVariable)(val position: InputPosition) extends EquivalenceModel {
```
Comment (sherfert) [inline community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala]: I think names that convey somehow the semantic meaning would be better instead of "TildeModel" and "AtModel"

File: community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala [hunk 3]
```diff
@@ -148,25 +148,38 @@ class InvalidNodePattern(
...
-                                properties: Option[Expression],
-                                direction: SemanticDirection,
-                                legacyTypeSeparator: Boolean = false
-                              )(val position: InputPosition) extends ASTNode {
+  variable: Option[LogicalVariable],
+  types: Seq[RelTypeName],
+  length: Option[Option[Range]],
+  properties: Option[Expression],
+  direction: SemanticDirection,
+  equivalenceModel: Option[EquivalenceModel] = None,
```
Comment (sherfert) [inline community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala]: What does it mean if you have no equivalence model? Or is that for backwards compatibility?

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/ExpandStarTest.scala [hunk 4]
```diff
@@ -60,77 +60,6 @@ class ExpandStarTest extends CypherFunSuite with AstConstructionTestSupport {
       "match p=(a:Start)-->(b) return a, b, p")
   }
 
-  test("rewrites * in RETURN GRAPHS") {
-    assertMultipleGraphsRewrite(
-      "WITH 1 AS a GRAPH AT 'url' AS foo, GRAPH AT 'url2' AS bar RETURN * GRAPHS *",
```
Comment (sherfert) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/ExpandStarTest.scala:65]: Do I understand correct that this syntax has been removed from the CIP? Or is there another reason theses tests were deleted?

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/normalizeGraphReturnItemsTest.scala [hunk 5]
```diff
@@ -21,54 +21,52 @@ package org.neo4j.cypher.internal.compiler.v3_4.ast.rewriters
...
-                Some(varFor("  FRESHID21")),
-                generated = true)(pos),None)(pos)))(pos),None,None,None,None)(pos),
-        Return(distinct = false,
-          ReturnItems(includeExisting = false,List(
-            UnaliasedReturnItem(SignedDecimalIntegerLiteral("1")(pos), "1")(pos)))(pos),None,None,None,None,Set())(pos)))(pos))(pos)
-
-    val result = original.rewrite(normalizeGraphReturnItems)
-    assert(result === expected)
-  }
+//  test("do not rename source graph") {
```
Comment (sherfert) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/normalizeGraphReturnItemsTest.scala]: We should either remove the code, if it is not needed any more, or make it work again.

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/Clause.scala [hunk 6]
```diff
@@ -77,85 +78,100 @@ sealed trait MultipleGraphClause extends Clause with SemanticAnalysisTooling {
...
-  override def name = "RELOCATE"
+  override def name = "RETURN GRAPH"
 
   override def semanticCheck: SemanticCheck =
     super.semanticCheck chain
-      graph.semanticCheck chain
       SemanticState.recordCurrentScope(this)
 }
 
+//final case class Relocate(graph: BoundGraphAs, to: GraphUrl)(val position: InputPosition)
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/Clause.scala]: More commented code...

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/createGraphIntroducesHorizon.scala [hunk 7]
```diff
@@ -14,39 +14,35 @@
...
-          )(p)
-          Seq(createGraph, newWith)
-
-        case clause =>
-          Some(clause)
-      }
-      query.copy(clauses = newClauses)(query.position)
-  })
-}
+//package org.neo4j.cypher.internal.frontend.v3_4.ast.rewriters
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/createGraphIntroducesHorizon.scala]: More commented code...

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/MultipleGraphClauseSemanticCheckingTest.scala [hunk 8]
```diff
@@ -35,16 +35,44 @@ class MultipleGraphClauseSemanticCheckingTest
...
-        |RETURN GRAPH result OF (a)""".stripMargin) shouldVerify { result: SemanticCheckResult =>
+      """CONSTRUCT { REMOVE a.prop }
+        |RETURN 1 as a
+      """.stripMargin) shouldVerify { result: SemanticCheckResult =>
+
+      result.errorMessages should equal(Set("Variable `a` not defined"))
+    }
+  }
 
+//  test("does not allow @ in CREATE") {
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/MultipleGraphClauseSemanticCheckingTest.scala]: More commented code...

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/MultipleGraphClauseSemanticCheckingTest.scala [hunk 9]
```diff
@@ -35,16 +35,44 @@ class MultipleGraphClauseSemanticCheckingTest
...
+
+  test("can conclude with RETURN GRAPH") {
+    parsing(
+      """RETURN GRAPH""".stripMargin) shouldVerify { result: SemanticCheckResult =>
       result.errors shouldBe empty
     }
   }
 
-  test("allows bound relationships in GRAPH OF") {
+  ignore("allows bound relationships in GRAPH OF") {
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/MultipleGraphClauseSemanticCheckingTest.scala]: These tests are ignored. Are they meant to be deleted or addresses in a different PR?

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/MultipleGraphClauseSemanticCheckingTest.scala [hunk 10]
```diff
@@ -1171,17 +1199,17 @@ class MultipleGraphClauseSemanticCheckingTest
...
   import scala.compat.Platform.EOL
 
-  private def contextsByPosition(state: SemanticState): String = {
-    val astNodes = state.recordedContextGraphs.keyPositionSet ++ state.recordedScopes.keyPositionSet
-    val keys = astNodes.toSeq.sortBy(x => x._2.offset -> x._1.toString)
-    val values = keys.map(ast => s"${state.recordedContextGraphs.getOrElse(ast._1, "--")}$EOL// $ast")
-    val finalContext = state.currentScope.contextGraphs.map(_.toString).getOrElse("--")
-    s"// Start$EOL${values.mkString(EOL)}$EOL$finalContext$EOL// End"
-  }
+  private def contextsByPosition(state: SemanticState): String = ??? //{
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/MultipleGraphClauseSemanticCheckingTest.scala]: ???

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphsParserTest.scala [hunk 11]
```diff
@@ -29,55 +29,52 @@ class MultipleGraphsParserTest
...
-        ast.GraphReturnItems(includeExisting = false, Seq(ast.NewContextGraphs(graphAt("foo", "url"))(pos)))(pos)
-      )(pos), ast.With(
-        ast.ReturnItems(includeExisting = false, Seq(ast.UnaliasedReturnItem(varFor("a"), "a")(pos)))(pos),
-        ast.GraphReturnItems(includeExisting = false, Seq(ast.ReturnedGraph(graph("foo"))(pos), ast.NewTargetGraph(graphAt("bar", "url2"))(pos)))(pos)
-      )(pos), ast.Return(
-        ast.GraphReturnItems(includeExisting = false, Seq(ast.ReturnedGraph(graph("bar"))(pos), ast.ReturnedGraph(graph("foo"))(pos)))(pos)
-      )(pos)
-    ))(pos)))
-  }
+//  test("WITH 1 AS a WITH a GRAPHS foo, >> GRAPH AT 'url2' AS bar RETURN GRAPHS bar, foo") {
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphsParserTest.scala]: More commented code...

**Review by sherfert — CHANGES_REQUESTED** [review]
This PR forgets to clean up a lot of commented code and ignored tests. Other than that it looks good to me.

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/semantics/SemanticFeature.scala (hunk 1 above)
Comment (Mats-SX) [reply community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/semantics/SemanticFeature.scala:24]: It depends. This was meant to separate the expression operator `~` that doesn't really have anything to do with Multiple Graphs, which could then be enabled independently whenever Neo4j Cypher starts to pick up Cypher 10 tasks.

File: community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala (hunk 2 above)
Comment (Mats-SX) [reply community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala]: I agree. I don't know what such names would be though.

File: community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala (hunk 3 above)
Comment (Mats-SX) [reply community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala]: It means that the user didn't write one in the pattern, like a normal pattern:

```
CREATE (n)     // no equiv model
CREATE (n~m)   // tilde equiv model
CREATE (n@m)   // at equiv model
```

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/ExpandStarTest.scala (hunk 4 above)
Comment (Mats-SX) [reply community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/ExpandStarTest.scala:65]: That's correctly understood.

File: community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala (hunk 2 above)
Comment (sherfert) [reply community/cypher/expressions-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/expressions/Pattern.scala]: Can you explain (in a short sentence) what each of these equivalence models mean? Maybe I can come up with good naming suggestions then

**Comment by Mats-SX** [comment]
@sherfert Thanks a lot for the review! We'll be updating this PR quite a lot, removing the ~ and @ parts from the patterns among other things.

We'll keep the `NOT READY FOR MERGE` on this until it is in a better shape.

**Comment by Mats-SX** [comment]
@sherfert This should be ready for review/merge now. The build was green prior to my rebase this morning.

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/Clause.scala [hunk 12]
```diff
@@ -95,9 +95,34 @@ final case class Clone(items: List[ReturnItem])(val position: InputPosition) ext
...
+    super.semanticCheck chain
+      SemanticPatternCheck.check(Pattern.SemanticContext.Create, pattern) chain
+      checkRelTypes(pattern)
+
+}
+
+trait SingleRelTypeCheck {
+  self: Clause =>
+
+  protected def checkRelTypes(pattern: Pattern): SemanticCheck =
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/Clause.scala]: Nice cleanup!

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphClausesParsingTest.scala [hunk 13]
```diff
@@ -17,104 +17,123 @@
...
-  private val nodePattern = exp.Pattern(List(exp.EveryPath(exp.NodePattern(None, List(), None)(pos))))(pos)
+  test("RETURN GRAPH") {
+    yields(ast.ReturnGraph(None))
+  }
 
-  private val complexPattern = exp.Pattern(List(
-    exp.NamedPatternPart(varFor("p"), exp.EveryPath(exp.NodePattern(None, List(), None)(pos)))(pos),
-    exp.NamedPatternPart(varFor("q"), exp.EveryPath(exp.NodePattern(None, List(), None)(pos)))(pos)
-  ))(pos)
+  // TODO: Parsing ambiguity; is it a graph name 'union' or no graph name and a UNION clause?
```
Comment (sherfert) [inline community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphClausesParsingTest.scala:135]: Here is a TODO left. When is it gonna be addressed?

**Review by sherfert — APPROVED** [review]
Looks good to me except there is one TODO left. Please look if it is something that needs resolving before this is merged or if it can be tracked and handled later.

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphClausesParsingTest.scala (hunk 13 above)
Comment (Mats-SX) [reply community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphClausesParsingTest.scala:135]: The TODO is actually part of a much bigger item; not allowing reserved words as symbolic names. That is, to not allow clause names such as `UNION` to be a variable. Once that is done, then this may be solved. (This extends to ugly things like `MATCH (match) WITH match.p AS with RETURN with AS return`)

Would you prefer that I remove the TODO from the code?

Comment (sherfert) [reply community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/parser/MultipleGraphClausesParsingTest.scala:135]: No, I just wanted to understand. But this is nothing your changes introduced. In that case it can stay there.
